dockstore-client/src/main/java/io/dockstore/client/cli/nested/WesLauncher.java

```diff
@@ -0,0 +1,200 @@
+package io.dockstore.client.cli.nested;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.List;
+
+import io.dockstore.openapi.client.ApiClient;
+import io.openapi.wes.client.model.RunId;
+import io.swagger.client.model.Workflow;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static io.dockstore.client.cli.ArgumentUtility.errorMessage;
+import static io.dockstore.client.cli.ArgumentUtility.out;
+import static io.dockstore.client.cli.Client.CLIENT_ERROR;
+
+public final class WesLauncher {
+
+    private static final Logger LOG = LoggerFactory.getLogger(WesLauncher.class);
+    private static final String TAGS = "{\"Client\":\"Dockstore\"}";
+    private static final String WORKFLOW_TYPE_VERSION = "1.0";
+    private static final String WORKFLOW_ENGINE_PARAMETERS = "{}";
+
+    private WesLauncher() {
+
+    }
+
+    /**
+     * This creates all the necessary parts of a WES launch, and then passes it on to the GA4GH WES client for the actual launch.
+     *
+     * @param workflowClient The WorkflowClient for the request
+     * @param workflowEntry The entry path, (i.e. github.com/myRepo/myWorkflow:version)
+     * @param workflowParamPath The path to a file to be used as an input JSON. (i.e. /path/to/file.json)
+     * @param attachments A list of paths to files to be attached to the request.
+     */
+    public static void launchWesCommand(WorkflowClient workflowClient, String workflowEntry, String workflowParamPath, List<String> attachments) {
+
+        // Get the workflow object associated with the provided entry path
+        final Workflow workflow = getWorkflowForEntry(workflowClient, workflowEntry);
+
+        // Can take the following values:
+        // 1. A TRS URL returning the raw primary descriptor file contents
+        // 2. TODO: A path to a file in the 'attachments' list
+        String workflowUrl = combineTrsUrlComponents(workflowClient, workflowEntry, workflow);
+
+        // A JSON object containing a key/value pair that points to the test parameter file in the 'attachments' list
+        // The key is WES server implementation specific. e.g. {"workflowInput":"params.json"}.
+        File workflowParams = loadFile(workflowParamPath);
+
+        // A list of supplementary files that are required to run the workflow. This may include any/all of the following:
+        // 1. The primary descriptor file
+        // 2. Secondary descriptor files
+        // 3. Test parameter files
+        // 4. Any other files referenced by the workflow
+        // TODO: Allow users to specify a directory to upload?
+        // TODO: Automatically attach all files referenced in remote Dockstore entry?
+        List<File> workflowAttachment = loadAttachments(attachments);
```

Comment:
Are the changes to workflow_params and workflow_attachment going to work on the other WES server implementations, e.g. Arvados, DNAStack, DNANexus, Illumina, etc.? Some of these WES implementations may no longer exist or may not be implemented correctly, but also we may not want to break compatibility with those that do exist? I recommend especially confirming that these changes are compatible with DNAStack and Patrick Magee <patrick@dnastack.com> and how he understands the schema. In addition there is a GA4GH ticket to remove the workflow_attachment from the WES schema [here](https://github.com/ga4gh/workflow-execution-service-schemas/issues/124) along with a discussion about alternatives. I advise trying to keep options open (not sure how exactly) since the schema seems to be in flux for quite some time now.

Reply:
Keep an eye on the issues and discussions [here](https://github.com/ga4gh/workflow-execution-service-schemas/issues)

Reply:
This actually wont work with DNAStack. We made a change a while back for TRS URLs that replaced the local file attachment approach. This ticket is to add back that functionality: https://ucsc-cgl.atlassian.net/browse/DOCK-1981.
We could potentially retarget these WES changes to a large feature branch until everything is ready to be moved to develop?

Thanks for pointing out the possible WES changes, I'll keep an eye on that.